### ISSUE TITLE
InitContainer execute multiple times till timeout error if "restartPolicy: Always"

### ISSUE THREAD

**Issue by masarlabs** [containers/podman#15631]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

When using a Kubernetes YAML file with a Pod definition using initContainers
and "restartPolicy: Always" the init container is executed multiple times till
a timeout error.

**Steps to reproduce the issue:**

1. This is the Kubernetes YAML file

```
# test.pod.yaml
apiVersion: v1
kind: Pod
metadata:
  name: test
  namespace: test
spec:
  restartPolicy: Always
  initContainers:
  - name: init
    image: busybox:stable
    command: [ "/bin/sh", "-c", "echo init $(date)" ]
  containers:
  - name: main
    image: busybox:stable
    command: [ "/bin/sh", "-c", "while true ; do echo main $(date) ; sleep 10 ; done" ]
```

2. Run podman with full debug, after 20 second of running there is the timeout error.

```
# podman play kube --log-level=debug test.pod.yaml
...
DEBU[0005] Created container 8b9041d4c86ee3ffd9d1038c491fa8f0ce5d8d622d74f8d3ce428b130485f79d in OCI runtime
DEBU[0005] Starting container 8b9041d4c86ee3ffd9d1038c491fa8f0ce5d8d622d74f8d3ce428b130485f79d with command [/bin/sh -c echo init $(date)] 
DEBU[0005] Started container 8b9041d4c86ee3ffd9d1038c491fa8f0ce5d8d622d74f8d3ce428b130485f79d

DEBU[0025] Exceeded conmon timeout waiting for container 8b9041d4c86ee3ffd9d1038c491fa8f0ce5d8d622d74f8d3ce428b130485f79d to exit

DEBU[0025] Removing container 8b9041d4c86ee3ffd9d1038c491fa8f0ce5d8d622d74f8d3ce428b130485f79d
DEBU[0025] Cleaning up container 8b9041d4c86ee3ffd9d1038c491fa8f0ce5d8d622d74f8d3ce428b130485f79d
DEBU[0025] Successfully cleaned up container 8b9041d4c86ee3ffd9d1038c491fa8f0ce5d8d622d74f8d3ce428b130485f79d
DEBU[0025] Unmounted container "8b9041d4c86ee3ffd9d1038c491fa8f0ce5d8d622d74f8d3ce428b130485f79d"
...
```

3. While podman is running, you can see from the logs that the init container
   is executed multiple times even if it terminates with 0 exit code.

```
# podman logs test-init

init Mon Sep 5 11:00:33 UTC 2022
init Mon Sep 5 11:00:33 UTC 2022
init Mon Sep 5 11:00:33 UTC 2022
init Mon Sep 5 11:00:34 UTC 2022
init Mon Sep 5 11:00:34 UTC 2022
init Mon Sep 5 11:00:34 UTC 2022
init Mon Sep 5 11:00:34 UTC 2022
init Mon Sep 5 11:00:34 UTC 2022
init Mon Sep 5 11:00:34 UTC 2022
init Mon Sep 5 11:00:34 UTC 2022
init Mon Sep 5 11:00:34 UTC 2022
init Mon Sep 5 11:00:35 UTC 2022
init Mon Sep 5 11:00:35 UTC 2022
```

**Describe the results you received:**

The init container is executed multiple times even if it terminates with 0 exit code.

**Describe the results you expected:**

The init container should be executed only one time.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
Client:       Podman Engine
Version:      4.2.0
API Version:  4.2.0
Go Version:   go1.19
Built:        Thu Jan  1 01:00:00 1970
OS/Arch:      linux/amd64
```

**Output of `podman info`:**

```
host:
  arch: amd64
  buildahVersion: 1.27.0
  cgroupControllers:
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon_2.1.3+ds1-1_amd64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.3, commit: unknown'
  cpuUtilization:
    idlePercent: 98.25
    systemPercent: 0.62
    userPercent: 1.14
  cpus: 8
  distribution:
    distribution: debian
    version: unknown
  eventLogger: journald
  hostname: masar
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 1000000001
      size: 999999
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 1000000001
      size: 999999
  kernel: 5.18.0-4-amd64
  linkmode: dynamic
  logDriver: journald
  memFree: 5263298560
  memTotal: 8226258944
  networkBackend: cni
  ociRuntime:
    name: crun
    package: crun_1.5+dfsg-1+b1_amd64
    path: /usr/bin/crun
    version: |-
      crun version 1.5
      commit: 54ebb8ca8bf7e6ddae2eb919f5b82d1d96863dea
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns_1.2.0-1_amd64
    version: |-
      slirp4netns version 1.2.0
      commit: 656041d45cfca7a4176f6b7eed9e4fe6c11e8383
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.4
  swapFree: 8998875136
  swapTotal: 8998875136
  uptime: 0h 42m 57.00s
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries: {}
store:
  configFile: /srv/masar/.config/containers/storage.conf
  containerStore:
    number: 2
    paused: 0
    running: 2
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /srv/masar/.local/share/containers/storage
  graphRootAllocated: 884717903872
  graphRootUsed: 654217465856
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /tmp/user/1000
  imageStore:
    number: 8
  runRoot: /run/user/1000/containers
  volumePath: /srv/masar/.local/share/containers/storage/volumes
version:
  APIVersion: 4.2.0
  Built: 0
  BuiltTime: Thu Jan  1 01:00:00 1970
  GitCommit: ""
  GoVersion: go1.19
  Os: linux
  OsArch: linux/amd64
  Version: 4.2.0
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman/now 4.2.0+ds1-3 amd64 [installed,local]
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/main/troubleshooting.md)**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

physical laptop with Debian GNU/Linux bookworm/sid

**Comment by rhatdan** [comment]
@umohnani8 @baude PTAL

**Comment by baude** [comment]
I have this pretty much nailed down.  I need to verify a few details with @mheon before I can cobble up a fix with any confidence.

**Comment by rhatdan** [comment]
@baude @mheon any update on this?

**Comment by rhatdan** [comment]
@baude @mheon any update on this?

**Comment by rhatdan** [comment]
This has been fixed in the main branch, should be fixed in podman 4.4.
